Index: bootstrap.inc
===================================================================
--- a/includes/bootstrap.inc	Tue Apr 06 01:00:52 2010 +0200
+++ b/includes/bootstrap.inc	Wed Apr 07 12:33:07 2010 +0200
@@ -1994,7 +1994,7 @@
   // the database is not yet initialized and we can't access any Drupal variables.
   // The file properties add more entropy not easily accessible to others.
   $filepath = DRUPAL_ROOT . '/includes/bootstrap.inc';
-  $key = sha1(serialize($databases) . filectime($filepath) . fileinode($filepath), TRUE);
+  $key = sha1(serialize($databases) /*. filectime($filepath) . fileinode($filepath)*/, TRUE);
   // The HMAC must match.
   return $hmac == base64_encode(hash_hmac('sha1', $check_string, $key, TRUE));
 }
@@ -2011,7 +2011,7 @@
     // check the HMAC before the database is initialized. filectime()
     // and fileinode() are not easily determined from remote.
     $filepath = DRUPAL_ROOT . '/includes/bootstrap.inc';
-    $key = sha1(serialize($databases) . filectime($filepath) . fileinode($filepath), TRUE);
+    $key = sha1(serialize($databases) /*. filectime($filepath) . fileinode($filepath)*/, TRUE);
   }
    // Generate a moderately secure HMAC based on the database credentials.
    $salt = uniqid('', TRUE);
